server/src/main/java/org/elasticsearch/search/rank/context/RankFeaturePhaseRankContext.java

```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.rank.context;
+
+public interface RankFeaturePhaseRankContext {}
```

Comment:
Why do we need this? I don't understand the requirement of the `abstract.. .implement empty interface`. Seems unnecessary? Why not make the abstract classes the base implementation?

Reply:
This was a suggestion on the WIP branch, so that there's an 1-1 association on the different contexes with exactly one search phase. The suggestion was to have them even as empty/just placeholders for now, and we can move up the hierarchy any common functionality. But I'm ok with removing these for now, and re-adding them iff needed.

Reply:
I believe @pmpailis is referring to a suggestion I made. There are four places now with a new phase that need to consider how to rank something. 

- query phase on the shard after the search(es) to collate results 
- end of the query phase on the coordinator to rank existing results as much as possible 
- rerank phase on the shard where we can collect information for a given re-ranker such as field data
- end of the rerank phase on the coordinator to use the newly collected information to rerank the original results

The idea is to have a different "context" for each place that needs to process rank in some way. This way as the builder is passed around and new information is known, we build a new context with that new information instead of updating and re-using the same one. For some rankers they may not need all of them and can just no-op them.

Reply:
> I believe @pmpailis is referring to a suggestion I made.

Tbh this came up during another sync that I had, but the idea is pretty similar :) The placeholder isn't "necessary", as for example `RankFeaturePhase` would know which set of contexes to use. It was just suggested as a nice thing to have, i.e.  a parent class to potentially hold all references to a single phase. Maybe there would be shared info along the way  (e.g. around profiling, but I'm not sure right now as I haven't got an impl for that yet) but we can remove these 2 references for now, and add them iff truly needed :)

Reply:
I suggested this in a sync where we were looking at the full implementation, and I gave two reasons. One is if we think of isolated search phases (query, fetch, rank etc), it makes sense to have a corresponding context object to each of those, so the reader can more easily follow the shape of the data the phase works, and also how the shard/coordinator sub-contexts are linked (through inheritance). The other reason is we can move some properties higher up in the abstraction - for example in the full implementation `windowSize` is common to both the coordinator/shard contexts for the query phase.